### PR TITLE
Database + Persistence Framework

### LINKED ISSUE #1: Persistence
**Summarize your proposal**
Storing data persistently is needed for nearly everything, therefore we need it too.

**Describe the solution you'd like**
The goal is to have a generic way to establish persistence for different components.
It has to provide a public interface for other components to store and load data without any knowledge about the technical implementation.

**Describe alternatives you've considered**
There are multiple possible technical implementation which have to be discussed.

### PR DESCRIPTION
**Pull request by Arkensor**
The database and persistence framework lays the foundation for all other gameplay systems in this framework so they can save and load data for individual players as well as across restarts. Closes #1 

### Documentation (read first and in order): 
- [Database Framework](https://github.com/Arkensor/EveronLife/blob/feature/persistence/docs/core/database/index.md)
- [Persistence Framework](https://github.com/Arkensor/EveronLife/blob/feature/persistence/docs/core/persistence/index.md)
- [New default loadout system](https://github.com/Arkensor/EveronLife/blob/feature/persistence/docs/feature/character/default_loadout.md)

### Included drivers
This system does not yet include any web API-based drivers. Their development is now starting, but in the meantime, gameplay content can be developed that will likely work out of the box with storage via web API into a proper relational database. The default driver right now is JSON local files, so you can read what is persisted during development. The default file driver for quick server setups in the future is intended as either binary files or a hosted web API the framework team crowdfunds. 

### Additional changes outside of the original issue scope
For the persistence to be testable some other foundations had to be laid down like a minimal player account system (where an account can have multiple characters) that is tied to the Bohemia UID that is unique per game purchase (not related to the BI account you might log into in-game). Also for new characters spawning a basic loadout system was added. This will be updated in the future into a full pre-game character selection + creation screen, for now, a player always has one character that is alive. Any killed characters become untied from a player's account and exist only as dead bodies in the world.

### Known issues
- Vehicle damage state is correctly persisted but not properly synced in MP for JIP replication. Nothing we can do about this. Need to wait for a fix on this: https://feedback.bistudio.com/T167370

### CODE REVIEW

**Comment by thebonbon** [comment]
Tested successfully with a simple job component on the player + action that lets you sign in / out.

Documentation was easy to understand. But I needed to find the right steps since the PR already has the world setup and the documentation starts from zero (Good for others to setup). Maybe add a little part for EveronLife only that just links to the steps we need? (eg. The PersistenceComponent and SaveData is already on the player etc..)


There are now 2 folders for Features (`Feature` and `Features`). Is this intended or can we merge them into one again?

**Comment by Arkensor** [comment]
> Documentation was easy to understand. But I needed to find the right steps since the PR already has the world setup and the documentation starts from zero (Good for others to setup). Maybe add a little part for EveronLife only that just links to the steps we need? (eg. The PersistenceComponent and SaveData is already on the player etc..)

I see, I will think about where to put a shortcut. Thank you for the feedback.

> There are now 2 folders for Features (`Feature` and `Features`). Is this intended or can we merge them into one again?


Yes, as per conventions the desired folder is `Feature` (singular). The old folder will be faded out by reworking all of the systems currently inside likely over the course of this month anyway.

**Review by thebonbon — COMMENTED** [review]
Static variables have different prefixes `m_` and `s_` or none.
Shouldn't they be identical?
Couple examples:
```
protected static int m_Sequence;
private static ref map<string, ref EL_DbContext> s_ContextCache;
protected static ref map<typename, typename> m_Mapping;
protected static ref EL_PlayerAccountManager s_Instance;
protected static const ResourceName PREFAB = "{B5D2C533815D511D}Prefabs/Core/EL_GlobalInventoryStorageManager.et";
protected static EL_GlobalInventoryStorageManagerComponent s_pInstance;
```

**Comment by Arkensor** [comment]
> protected static int m_Sequence;
private static ref map<string, ref EL_DbContext> s_ContextCache;
protected static ref map<typename, typename> m_Mapping;
protected static ref EL_PlayerAccountManager s_Instance;

Those were indeed mismatches, I overlooked them. Fixed

> protected static const ResourceName PREFAB = "{B5D2C533815D511D}Prefabs/Core/EL_GlobalInventoryStorageManager.et";

That one is correct because even if static the `const` modified takes priority as per AR conventions so it must be all upper case.

File: src/Scripts/Game/Core/Database/Drivers/EL_DbDriverRegistry.c [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+			{
+				EL_DbDriverRegistry.Register(alias, driverType);
+			}
+		}
+	}
+}
+
+class EL_DbDriverRegistry
+{
+	protected static ref map<string, typename> m_Drivers;
```
Comment (thebonbon) [inline src/Scripts/Game/Core/Database/Drivers/EL_DbDriverRegistry.c:20]:
```suggestion
	protected static ref map<string, typename> s_Drivers;
```

File: src/Scripts/Game/Core/Database/Drivers/InMemory/EL_InMemoryDbDriver.c [hunk 2]
```diff
@@ -0,0 +1,148 @@
+[EL_DbDriverName(EL_InMemoryDbDriver, {"InMemory"})]
+class EL_InMemoryDbDriver : EL_DbDriver
+{
+	protected static ref map<string, ref EL_InMemoryDatabase> m_Databases;
```
Comment (thebonbon) [inline src/Scripts/Game/Core/Database/Drivers/InMemory/EL_InMemoryDbDriver.c:4]:
```suggestion
	protected static ref map<string, ref EL_InMemoryDatabase> s_Databases;
```

File: src/Scripts/Game/Core/Database/EL_DbEntity.c [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	//! Check if an id has been assigned
+	bool HasId()
+	{
+		return m_sId;
+	}
+}
+
+class EL_DbName
+{
+	protected static ref map<typename, string> m_mMapping;
```
Comment (thebonbon) [inline src/Scripts/Game/Core/Database/EL_DbEntity.c:32]:
```suggestion
	protected static ref map<typename, string> s_mMapping;
```

File: src/Scripts/Game/Core/Persistence/EL_PersistentScriptedStateBase.c [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+	bool ApplyTo(notnull EL_PersistentScriptedStateBase scriptedState)
+	{
+		return EL_DbEntityUtils.StructAutoCopy(this, scriptedState);
+	}
+}
+
+class EL_PersistentScriptedStateSettings
+{
+	protected static ref map<typename, ref EL_PersistentScriptedStateSettings> m_Settings;
+	protected static ref map<typename, typename> m_ReverseMapping;
```
Comment (thebonbon) [inline src/Scripts/Game/Core/Persistence/EL_PersistentScriptedStateBase.c:154]:
```suggestion
	protected static ref map<typename, ref EL_PersistentScriptedStateSettings> s_mSettings;
	protected static ref map<typename, typename> s_mReverseMapping;
```

File: src/Scripts/Game/Core/Persistence/EL_PersistentWorldEntityLoader.c [hunk 5]
```diff
@@ -0,0 +1,228 @@
+class EL_PersistentWorldEntityLoader
+{
+	protected static ref map<string, typename> m_SaveDataTypeCache;
```
Comment (thebonbon) [inline src/Scripts/Game/Core/Persistence/EL_PersistentWorldEntityLoader.c:3]:
```suggestion
	protected static ref map<string, typename> s_mSaveDataTypeCache;
```

**Review by thebonbon — CHANGES_REQUESTED** [review]

**Review by thebonbon — APPROVED** [review]

**Pull request merged**